x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -200,20 +206,50 @@ public void updateResponse(String docId,
                             R response,
                             ActionListener<UpdateResponse> listener) {
         try {
-            Map<String, Object> source = new HashMap<>();
-            source.put(RESPONSE_HEADERS_FIELD, responseHeaders);
-            source.put(RESULT_FIELD, encodeResponse(response));
+            final XContentBuilder source = buildSourceForUpdateRequest(response, Map.of(RESPONSE_HEADERS_FIELD, responseHeaders));
             UpdateRequest request = new UpdateRequest()
                 .index(index)
                 .id(docId)
-                .doc(source, XContentType.JSON)
+                .doc(source)
                 .retryOnConflict(5);
             clientWithOrigin.update(request, listener);
         } catch(Exception e) {
             listener.onFailure(e);
         }
     }
 
+    XContentBuilder buildSourceForUpdateRequest(R response, Map<String, Object> otherFields) throws IOException {
+        final BytesStreamOutput buffer = new BytesStreamOutput(); // TODO: Integrate with the circuit breaker
+        XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON, buffer);
+        builder.startObject();
+        for (Map.Entry<String, Object> e : otherFields.entrySet()) {
+            if (e.getKey().equals(RESULT_FIELD) == false) {
+                builder.field(e.getKey(), e.getValue());
+            }
+        }
+        // Write the base64 encoding directly to OutputStream of XContentBuilder avoid using multiple buffers.
+        final XContentGenerator generator = builder.generator();
+        if (generator instanceof JsonXContentGenerator == false) {
+            assert false : "expected a JsonXContentGenerator; got " + generator;
+            throw new IllegalStateException("expected a JsonXContentGenerator; got " + generator);
+        }
+        final JsonXContentGenerator jsonGenerator = (JsonXContentGenerator) generator;
+        final OutputStream out = builder.getOutputStream();
+        builder.field(RESULT_FIELD);
+        builder.flush();
+        out.write(':');
+        out.write('\"');
```

Comment:
@nik9000 Do you have a better/cleaner way to do this?

Reply:
Looks like we have a `writeRawField` but it wants an `InputStream` rather than gives you an output stream. I wonder if we could add a variant that takes a `Consumer<OutputStream>`. That feels like it might be a little cleaner, but you'd have to build it.

Reply:
I've opened https://github.com/elastic/elasticsearch/pull/73804.